Filter non input symexprs from codecache guards

oulgen opened the pull request:
Summary: Dynamo lifts all symexprs that appear in the inputs to top level which means that we do not need to look at guards that contain symexprs that do not appear in the inputs. Prune them.

Test Plan: added two new tests

Differential Revision: D58200476




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -4232,6 +4232,17 @@ def evaluate_guards_for_args(self, placeholders, args, *, ignore_static=True):
+        """
+        Removes guards that refer to non input symints
+        """
+        symints = {str(s) for s in symints}
```
There's no reason to stringify these for the set? Symbols are hashable just fine? (Actually, is symints SymInt or sympy.Symbol?)

oulgen replied:
`symints` are torch.SymInt; while, `self.guards[0].expr.free_symbols` are `sympy.core.symbol.Symbol`. I assume I need to convert them both to strings to compare them?

ezyang replied:
Instead of what you're doing here, do `s.node.expr for s in symints if isinstance(s, torch.SymInt) and isinstance(s.node.expr, sympy.Symbol)`. Drop the SymInt test if you can guarantee these are SymInts. I can't actually tell from context.

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -904,6 +905,7 @@ def _save_graph(
         shape_env = FxGraphCache._get_shape_env()
         assert shape_env is not None
         symints = FxGraphCache._filter_backed_symints(example_inputs)
```
So, what's the plan for unbacked symints allocated in forwards that end up as inputs to the backwards graph?

oulgen replied:
On https://github.com/pytorch/pytorch/pull/126251 @masnesral claims that unbacked symints cant affect guards

bdhirsh replied:
Curiosity question: can runtime asserts be associated with the backward graph today? Or do they only show up in forward (if they show up in backward, do include them directly inside the backward graph as fx nodes?)

bdhirsh commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -4232,6 +4232,17 @@ def evaluate_guards_for_args(self, placeholders, args, *, ignore_static=True):
+        for g in self.guards:
+            if all(str(s) in symints for s in g.expr.free_symbols):
+                guards.append(g)
+        self.guards = guards
```
Hmmm it's not clear to me why this is correct. Say we have this code:

```
@torch.compile
def f(x, val):
    if val > 5:
        return x.sin()
    else:
        return x.cos()

x = torch.randn(2)
out1 = f(x, 6)
out2 = f(x, 4)
```

We need to dynamo to fail a guard for `s0 > 5` on the second invocation of `f`. But... if I understand correctly, won't this change in-place mutate the ShapeEnv and delete that guard (while inductor is compiling), before dynamo has a chance to grab all relevant guards from the ShapeEnv and codegen them?

oulgen replied:
@bdhirsh and i looked at this example with dynamic=True and it appears to not fit the cache at all and ends up in a fresh compile

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -772,6 +772,7 @@ def _lookup_graph(
         assert shape_env is not None
 
         symints = FxGraphCache._filter_backed_symints(example_inputs)
+        shape_env.drop_non_input_symint_guards(symints)
```
@ezyang can you verify its safe to manipulate the existing shape env? I high-level understanding from @bdhirsh is that dynamo will still evaluate these guards after inductor is finished doing its thing. If it is safe to remove this category of guards, why don't we remove them unconditionally before we start inducting rather than putting the manipulation in the cache path?

ezyang replied:
No, you definitely cannot start dropping guards from the shape env lol

oulgen replied:
Fixed this on a previous version of the PR

oulgen commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2463,6 +2463,7 @@ def _init(
         )
 
         self.guards: List[ShapeGuard] = []
+        self.input_guards: List[ShapeGuard] = []
```
i ended up separating these because symbolic shape recording pass validation was failing otherwise

ezyang replied:
cc @ysiraichi

ezyang replied:
This seems like a dorky way to do this. One superficial complaint is that ShapeEnv is not Inductor specific, and yet we have named this thing "input guards", what the heck is an input guard?

For me, an intuitive code organization is that I want some way to call `produce_guards` with an explicit guards list, rather than using the guards that is accumulated directly on the ShapeEnv. That seems like a very modest change and doesn't require adding an extra field on ShapeEnv.

You need to make sure to not run translation validation on this `produce_guards` call either, as it will fail (and for good reason). TV should only run when you don't explicitly pass in guards to produce.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -4240,6 +4254,19 @@ def evaluate_guards_for_args(self, placeholders, args, *, ignore_static=True):
             return self.evaluate_guards_expression(code, args)
         return True
 
+    def get_input_guards(self, symints):
```
Nit: don't call this `get_input_guards`. Maybe something like `get_pruned_guards`, doc'ed as "Get a list of guards, but pruned so it only provides guards that reference symints from the passed in input." TBH, it would also be more logical for this to take a set of symbols instead of a list of torch.SymInt. In any case, a type annotation helps.

ezyang reviewed (approved):
This seems substantively fine.

masnesral reviewed (approved):
Looks correct based on discussions, but I still don't understand everything about guards so I'll defer to @ezyang about the overall soundness.

One thing that makes me nervous is that we're using a different set of guards for the cache than elsewhere. So if we've made a mistake and these guards are too lax, it's going to be very confusing and hard to debug. It would seem better, for example, if we could distinguish those guards added by inductor -- then use them directly in the cache rather than filtering from this single list based on criteria that I'm not sure are codified. But I presume that distinguishing the inductor-added guards would be a large refactor?

oulgen committed:
Filter non input symexprs from codecache guards (#128052)

Summary:

Dynamo lifts all symexprs that appear in the inputs to top level which means that we do not need to look at guards that contain symexprs that do not appear in the inputs. Prune them.

Test Plan: added two new tests

Reviewed By: masnesral

Differential Revision: D58200476
